### PR TITLE
Spacecat 672 benchmark 4000 pi digits

### PR DESCRIPTION
**Pull request by tsmado**
WIP, need to resolve how to log the test results.

### CODE REVIEW

File: cloud_level_testing/features/steps/definitions.py [hunk 1]
```diff
@@ -601,3 +605,29 @@ def attach_floating_ip_to_server(context, server_name):
...
         ip = context.client.add_auto_ip(server=server, wait=True)
+        context.vm_ip_address = ip
+        context.logger.log_info(f"Attached floating ip: {ip}")
+    
+    @then('I start calculating 4000 digits of pi on VM and check the ping response')
+    def calculate_pi_on_vm(context):
+        """
+        """
+        # config
+        ping_retries = 60
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/definitions.py:616]: can we move that to some class constant?

File: cloud_level_testing/features/steps/definitions.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import string
 
 from openstack.exceptions import DuplicateResource
-from libs.ConnectivityClient import SshClient
+from libs.ConnectivityClient import SshClient, run_async_command
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/definitions.py:9]: are you still using async?

File: cloud_level_testing/features/steps/tools.py [hunk 3]
```diff
@@ -181,62 +185,65 @@ def collect_ips(client,logger:Logger):
         assertline = f"No ips found"
     return ips, assertline
 
-def collect_jhs(client, test_name,logger:Logger):
+
+def collect_jhs(client, test_name, logger: Logger):
     servers = client.compute.servers()
-    lookup = test_name+"-jh"
-    lookup = "default-jh" # just for testing delete later
+    lookup = test_name + "-jh"
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/tools.py:191]: please use fstring 

lookup = f"{test_name}-jh"

File: cloud_level_testing/features/steps/tools.py [hunk 4]
```diff
@@ -181,62 +185,65 @@ def collect_ips(client,logger:Logger):
...
+    lookup = "default-jh"  # just for testing delete later
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/tools.py:192]: why are you defining lookup two times?

File: cloud_level_testing/features/steps/tools.py [hunk 5]
```diff
@@ -284,29 +291,33 @@ def create_security_group_rule(
...
             if context.client.delete_server(vm_id, wait=True):
                 context.logger.log_info(f"VM {vm_id} deleted")
-                context.collector.virtual_machines.remove(vm_id)
+                try:
+                    context.collector.virtual_machines.remove(vm_id)
+                except ValueError:
+                    if vm_id in context.collector.jumphosts:
+                        context.collector.jumphosts.remove(vm_id)
             else:
                 context.logger.log_info(f"FAIL! VM {vm_id} wasn't deleted")
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/tools.py:312]: please rephrase the message without "FAIL!"

File: cloud_level_testing/features/steps/tools.py [hunk 6]
```diff
@@ -373,14 +384,92 @@ def delete_ports(context, port_ids: list = None):
                 context.logger.log_info(f"FAIL! Port {port_id} wasn't deleted")
 
 
+def delete_jumphosts(context, jumphost_ids: list = None):
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/tools.py:387]: default parameters without spaces

File: cloud_level_testing/features/steps/tools.py [hunk 7]
```diff
@@ -373,14 +384,92 @@ def delete_ports(context, port_ids: list = None):
...
+
+
+def calc_average(values: list[float]) -> float:
+    sum = 0
+    for value in values:
+        sum = sum + value
+    return sum / len(values)
+
+
+def get_timestamps(data: str) -> tuple[str, str]:
```
Comment (piobig2871) [inline cloud_level_testing/features/steps/tools.py:445]: can you please let me know where the actual timestamps are from?

File: libs/ConnectivityClient.py [hunk 8]
```diff
@@ -1,28 +1,37 @@
...
 class MetricLabels:
-    STATUS_CODE = 'status_code'
-    HOST = 'host'
-    ENDPOINT = 'endpoint'
+    STATUS_CODE = "status_code"
+    HOST = "host"
+    ENDPOINT = "endpoint"
+
 
 class ResultStatusCodes:
```
Comment (piobig2871) [inline libs/ConnectivityClient.py:17]: you can use the http lib for that 

example:
from http import HTTPStatus

HTTPStatus.OK == 200
>>> True

There is no flag as success afaik

File: libs/ConnectivityClient.py [hunk 9]
```diff
@@ -124,9 +147,10 @@ def test_internet_connectivity(self, conn_test, ip='8.8.8.8', tot_ips=1):
             Raises:
                 Assertion Failed: Failed to test internet connectivity for endpoint, if IP address is in wrong format or unreachable
         """
-        self.assertline=""
+        self.assertline = ""
```
Comment (piobig2871) [inline libs/ConnectivityClient.py:150]: not used

File: libs/ConnectivityClient.py [hunk 10]
```diff
@@ -207,4 +231,40 @@ def print_working_directory(self):
...
+            self.connect()
+            self.close_conn()
+            return True
+        except (
+            paramiko.ssh_exception.NoValidConnectionsError,
+            paramiko.ssh_exception.SSHException,
+        ):
+            return False
+        except Exception as e:
+            print(f"Error occurred: {e}")
```
Comment (piobig2871) [inline libs/ConnectivityClient.py:251]: change to log please

File: cloud_level_testing/features/steps/tools.py (hunk 3 above)
Comment (tsmado) [reply cloud_level_testing/features/steps/tools.py:191]: This is not from made by me but i will change it.

File: cloud_level_testing/features/steps/tools.py (hunk 4 above)
Comment (tsmado) [reply cloud_level_testing/features/steps/tools.py:192]: Again, its something Kata uses i think. @fraugabel

File: libs/ConnectivityClient.py (hunk 8 above)
Comment (tsmado) [reply libs/ConnectivityClient.py:17]: Again, @fraugabel can probably elaborate as this is something that came with rebase.

File: libs/ConnectivityClient.py (hunk 9 above)
Comment (tsmado) [reply libs/ConnectivityClient.py:150]: @fraugabel

File: libs/ConnectivityClient.py (hunk 10 above)
Comment (tsmado) [reply libs/ConnectivityClient.py:251]: Done.

File: cloud_level_testing/features/steps/tools.py (hunk 7 above)
Comment (tsmado) [reply cloud_level_testing/features/steps/tools.py:445]: Timestamps are from the VMs, they are taken before and after the calculation command.

File: cloud_level_testing/features/steps/definitions.py (hunk 2 above)
Comment (tsmado) [reply cloud_level_testing/features/steps/definitions.py:9]: No forgot to remove the unused import.

File: cloud_level_testing/features/steps/definitions.py (hunk 1 above)
Comment (tsmado) [reply cloud_level_testing/features/steps/definitions.py:616]: Do you wanna have it defined in the features or just a constant?

Comment (piobig2871) [reply cloud_level_testing/features/steps/definitions.py:616]: just a constant

**Pull request merged**